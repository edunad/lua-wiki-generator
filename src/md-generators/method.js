#!/usr/bin/env node
'use strict';

class MDMethod {
    #outputFolder;
    #template;

    constructor(outputFolder, template) {
        this.#outputFolder = outputFolder;
        this.#template = template;
    }

    generate = (data) => {
        if (!this.#template) return '';
        let fixedTemplate = this.#template;

        fixedTemplate = this.#parseDescription(fixedTemplate, data);
        fixedTemplate = this.#parseScope(fixedTemplate, data);
        fixedTemplate = this.#parseTitle(fixedTemplate, data);
        fixedTemplate = this.#parseMethodName(fixedTemplate, data);
        fixedTemplate = this.#parseHint(fixedTemplate, data);
        fixedTemplate = this.#parseParameters(fixedTemplate, data);
        fixedTemplate = this.#parseReturns(fixedTemplate, data);
        fixedTemplate = this.#parseExamples(fixedTemplate, data);

        return fixedTemplate;
    };

    #parseDescription = (template, data) => {
        if (!data.commentBlock.description) return template;
        return template.replace('$DESCRIPTION$', `${data.commentBlock.description}\n`);
    };

    #parseScope = (template, data) => {
        if (!data.commentBlock.env) return template;
        return template.replace(/\$SCOPE\$/g, data.commentBlock.env.toLowerCase());
    };

    #parseTitle = (template, data) => {
        if (!data.title) return template;

        return template.replace(
            '$TITLE_METHOD_NAME$',
            data.title.link
                ? `[${data.title.link}](${this.#outputFolder}/${data.title.link})${data.title.msg.replace(data.title.link, '')}`
                : data.title.msg,
        );
    };

    #parseMethodName = (template, data) => {
        if (!data.method) return template;
        return template.replace('$METHOD_NAME$', data.method);
    };

    #parseHint = (template, data) => {
        let hints = '';
        if (data.commentBlock.hints.length > 0) {
            hints += '\n';

            data.commentBlock.hints.forEach((hint) => {
                if (hint.message.trim() === '') return;
                hints += `{% hint style="${hint.type}" %} ${hint.message} {% endhint %}\n`;
            });

            hints += '\n';
        }

        return template.replace('$HINTS$', hints);
    };

    #parseParameters = (template, data) => {
        let params = '';
        if (data.commentBlock.params.length > 0) {
            params += `------\n`;
            params += `## Parameters\n\n`;
            params += `| Type   | Name | Description | Optional |\n`;
            params += `| ------ | ---- | ----------- | -------: |\n`;

            data.commentBlock.params.forEach((param) => {
                const type = param.link ? `[${param.link}](${this.#outputFolder}/${param.link})` : param.type;
                params += `| ${type} | ${param.name} | ${param.description} | ${param.optional ? 'âœ”' : ''} |\n`;
            });

            params += '\n';
        }

        return template.replace('$PARAMETERS$', params);
    };

    #parseReturns = (template, data) => {
        let returns = '';
        if (data.commentBlock.returns.length > 0) {
            returns += `------\n`;
            returns += `## Returns\n\n`;
            returns += `| Type   | Description |\n`;
            returns += `| ------ | ----------: |\n`;

            data.commentBlock.returns.forEach((ret) => {
                const type = ret.link ? `[${ret.link}](${this.#outputFolder}/${ret.link})` : ret.type;
                returns += `| ${type} | ${ret.description} |\n`;
            });

            returns += '\n';
        }

        return template.replace('$RETURNS$', returns);
    };

    #parseExamples = (template, data) => {
        let example = '';
        if (data.commentBlock.examples.length > 0) {
            example += '```lua\n';
            data.commentBlock.examples.forEach((ex) => {
                example += `${ex}\n`;
            });
            example += '```\n\n';
        }

        return template.replace('$EXAMPLE$', example);
    };
}

module.exports = MDMethod;
