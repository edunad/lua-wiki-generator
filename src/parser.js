#!/usr/bin/env node
'use strict';

import fs from 'fs';
import readline from 'readline';
import events from 'events';

export class LuaParser {
    #currentCommentBlock;
    #currentMode;
    #parsedBlocks;
    #filePath;
    #foundBlock;

    constructor(path) {
        this.#filePath = path;
    }

    parseLua = async () => {
        const stream = fs.createReadStream(this.#filePath, 'utf8');
        const rl = readline.createInterface({
            input: stream,
            crlfDelay: Infinity,
        });

        // Reset
        this.#reset();

        // Parse
        rl.on('line', (line) => this.#parseLine(line));
        await events.once(rl, 'close');

        // Finish
        if (this.#parsedBlocks.length <= 0) console.warn(`File ${this.#filePath} has no lua documentation?`);
        return Promise.resolve(this.#parsedBlocks);
    };

    #reset = () => {
        this.#currentCommentBlock = this.#getDefaultBlock();
        this.#currentMode = 'NONE';
        this.#parsedBlocks = {
            file: this.#filePath,
            blocks: [],
        };
    };

    #isIgnoredAttr = (attr) => {
        return ['---@meta', '---@class'].indexOf(attr) !== -1; // TODO: Add other weird attributes that we don't care
    };

    #isExtensionMethod = (line) => {
        return line.indexOf('function(') !== -1;
    };

    #isClass = (line) => {
        return line.indexOf('={') !== -1;
    };

    #isMethod = (line) => {
        return line.startsWith('function ');
    };
    // ------

    #extractMethod = (line) => {
        return [...line.matchAll(/function ((.*):.*)\((.*)\)/g)][0];
    };

    #extractExtension = (line) => {
        return [...line.matchAll(/(.*)=.*function.*(\(.*\))/g)][0];
    };

    #extractClass = (line) => {
        return [...line.matchAll(/(.*)=.*{/g)][0];
    };

    // DOC PARSING ----
    #parseParam = (line) => {
        //---@param callback function bla
        const paramRegex = [...line.matchAll(/---@param ([^ \r\n]*) ([^ \r\n]*) ?"?([^"\r\n]*)?"?/g)];
        if (!paramRegex || paramRegex.length !== 1) return null;

        const params = paramRegex[0];
        if (!params || params.length < 3) return null;
        const isOptional = params[1].indexOf('?') !== -1;

        return {
            name: params[1].replace('?', ''),
            type: params[2],
            description: params[3] || 'No description',
            optional: isOptional,

            link: this.#isLuaPrimitive(params[2]) ? null : params[2],
        };
    };

    #parseField = (line) => {
        //---@field callback function bla
        const fieldRegex = [...line.matchAll(/---@field ([^ \r\n]*) ([^ \r\n]*) ?"?([^"\r\n]*)?"?/g)];
        if (!fieldRegex || fieldRegex.length !== 1) return null;

        const fields = fieldRegex[0];
        if (!fields || fields.length < 2) return null;

        return {
            name: fields[1],
            type: fields[2],

            link: this.#isLuaPrimitive(fields[2]) ? null : fields[2],
        };
    };

    #parseReturn = (line) => {
        //---@return boolean asdasdas
        const returnRegex = [...line.matchAll(/---@return ([^ \r\n]*) ?"?([^"\r\n]*)?"?/g)];
        if (!returnRegex || returnRegex.length !== 1) return null;

        const returns = returnRegex[0];
        if (!returns || returns.length <= 0) return null;

        return {
            type: returns[1],
            description: returns[2] || 'No description',

            link: this.#isLuaPrimitive(returns[1]) ? null : returns[1],
        };
    };

    #parseHints = (line) => {
        //---@hint @warning PAGE / FUNCTIONALITY STILL IN CONSTRUCTION
        const hintRegex = [...line.matchAll(/---@hint ([^ \r\n]*) ?"?([^"\r\n]*)?"?/g)];
        if (!hintRegex || hintRegex.length !== 1) return null;

        const hints = hintRegex[0];
        if (!hints || hints.length !== 3) return null;

        return {
            type: hints[1],
            message: hints[2],
        };
    };

    #parseEnv = (line) => {
        //---@env shared
        const envRegex = [...line.matchAll(/---@env ([^ \r\n]*)/g)];
        if (!envRegex || envRegex.length !== 1) return 'UNKNOWN';

        const envs = envRegex[0];
        if (!envs || envs.length !== 2) return 'UNKNOWN';

        return envs[1];
    };
    /// --------

    #buildReturn = (returns) => {
        if (!returns || returns.length <= 0) return 'void';
        return returns.map((ret) => ret.type).join(', ');
    };

    /// --------
    #getDefaultBlock = () => {
        return {
            env: 'UNKNOWN',

            params: [],
            hints: [],
            returns: [],
            fields: [],
            description: [],
            examples: [],
        };
    };

    #isLuaPrimitive = (type) => {
        const primitives = ['nil', 'boolean', 'number', 'string', 'function', 'userdata', 'thread', 'table'];

        const fixedStr = type.toLowerCase().trim();
        const name = fixedStr.split('[');

        return primitives.includes(name[0] || fixedStr);
    };

    // ------------
    #parseLine = (line) => {
        const trimmedLine = line.replace(/\s/g, '');
        if (trimmedLine === '' || this.#isIgnoredAttr(trimmedLine)) return; // Ignore

        // Reached a method?
        if (!trimmedLine.startsWith('---')) {
            if (this.#foundBlock) {
                if (this.#isMethod(line)) {
                    // ex: function console:execute(args) end
                    const methodData = this.#extractMethod(line);
                    if (methodData && methodData.length >= 3) {
                        this.#parsedBlocks.blocks.push({
                            type: 'METHOD',
                            title: {
                                msg: methodData[1].trim(),
                                link: this.#isLuaPrimitive(methodData[2]) ? null : methodData[2],
                            },
                            method: methodData[0].replace('function', this.#buildReturn(this.#currentCommentBlock.returns)),
                            commentBlock: this.#currentCommentBlock,
                        });
                    }
                } else if (this.#isExtensionMethod(trimmedLine)) {
                    const methodData = this.#extractExtension(line);
                    if (methodData && methodData.length >= 3) {
                        this.#parsedBlocks.blocks.push({
                            type: 'EXTENSION',
                            title: {
                                msg: methodData[1].trim(),
                                link: null,
                            },
                            method: `${this.#buildReturn(this.#currentCommentBlock.returns)} ${methodData[1].trim()}${methodData[2]}`,
                            commentBlock: this.#currentCommentBlock,
                        });
                    }
                } else if (this.#isClass(trimmedLine)) {
                    const methodData = this.#extractClass(line);
                    if (methodData && methodData.length >= 1) {
                        this.#parsedBlocks.blocks.push({
                            type: 'CLASS',
                            title: {
                                msg: methodData[1].trim(),
                                link: null,
                            },
                            method: '',
                            commentBlock: this.#currentCommentBlock,
                        });
                    }
                }

                this.#foundBlock = false;
            }

            this.#currentCommentBlock = this.#getDefaultBlock(); // Reset
        } else {
            if (line.startsWith('---@param')) {
                const param = this.#parseParam(line);
                if (param) this.#currentCommentBlock.params.push(param);
            } else if (line.startsWith('---@return')) {
                const ret = this.#parseReturn(line);
                if (ret) this.#currentCommentBlock.returns.push(ret);
            } else if (line.startsWith('---@field')) {
                const ret = this.#parseField(line);
                if (ret) this.#currentCommentBlock.fields.push(ret);
            } else if (line.startsWith('---@env')) {
                this.#currentCommentBlock.env = this.#parseEnv(line);
            } else if (line.startsWith('---@hint')) {
                const ret = this.#parseHints(line);
                if (ret) this.#currentCommentBlock.hints.push(ret);
            } else if (line.startsWith('---*')) {
                this.#currentCommentBlock.description.push(line.replace('---*', '').trim());
            } else if (line.startsWith('---```lua')) {
                this.#currentMode = 'EXAMPLE';
            } else if (line.startsWith('---```')) {
                this.#currentMode = 'NONE';
            } else if (this.#currentMode === 'EXAMPLE') {
                this.#currentCommentBlock.examples.push(line.replace('---', '').trim());
            }

            this.#foundBlock = true;
        }
    };
}
