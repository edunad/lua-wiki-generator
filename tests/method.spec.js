#!/usr/bin/env node
'use strict';

import fs from 'fs-extra';
import { fileURLToPath } from 'url';
import { dirname, resolve } from 'path';

import { MDMethod } from '../src/md-generators/method.js';
import { LuaParser } from '../src/parser.js';

const __filename = fileURLToPath(import.meta.url);
const __dirname = dirname(__filename);

const METHOD_TEMPLATE_FILE = './__test_templates__/METHOD_TEMPLATE.md';

describe('MDGenerators', () => {
    it('Generates a Method MD', async () => {
        const methodTemplate = await fs.readFile(resolve(__dirname, METHOD_TEMPLATE_FILE), 'utf8');

        const md = new MDMethod('./home', methodTemplate);
        expect(md).not.toBe(undefined);

        const parser = new LuaParser(resolve(__dirname, `./ias.method_test.lua`));
        expect(parser).not.toBe(undefined);

        const data = await parser.parseLua();
        expect(data).not.toBe(undefined);
        expect(data.blocks).not.toBe(undefined);
        expect(data.blocks.length).toBe(1);

        const mdGen = md.generate(data.blocks[0]);
        expect(mdGen).toMatchSnapshot();
    });
});
