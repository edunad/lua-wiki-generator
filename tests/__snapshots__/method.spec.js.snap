// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`MDGenerators Generates a Method MD 1`] = `
"## ![shared](.gitbook/assets/shared.png) [console](./home/console)

{% hint style="@warning" %} PAGE / FUNCTIONALITY STILL IN CONSTRUCTION {% endhint %}
{% hint style="@danger" %} OH NOOOOOOO {% endhint %}


\`\`\`lua
boolean, string console:execute(args)
\`\`\`

Executes a console command programmatically
\`\`\`lua
--my_command is the command to execute, true is a parameter
console:execute({"my_command", "true"})
\`\`\`

------
## Parameters

| Type   | Name | Description | Optional |
| ------ | ---- | ----------- | -------: |
| string[] | args | No description |  |

------
## Returns

| Type   | Description |
| ------ | ----------: |
| boolean | If command was executed successfully |
| string | No description |

"
`;
